File: DynamicProgramming/PalindromePartitioningII.cpp
#include <iostream>
#include <string>

using std::string;

class Solution
{
public:
    int minCut(const string& s) {
        int lens = s.size();
        int dp[lens+1];
        bool P[lens][lens];

        for (int i = 0; i <= lens; i++)
            dp[i] = lens - i - 1;

        for (int i = 0; i < lens; i++) {
            for (int j = 0; j < lens; j++)
                P[i][j] = false;
        }

        for (int i = lens - 1; i >= 0; i--) {
            for (int j = i; j < lens; j++) {
                if (s[i] == s[j] && (j - i <= 1 || P[i+1][j-1])) {
                    P[i][j] = true;
                    dp[i] = std::min(dp[i], dp[j+1]+1);
                }
            }
        }
        return dp[0];
    }
};

int main()
{
    string str ("abcbabaa");
    Solution s;
    std::cout << s.minCut(str) << "\n";
    return 0;
}

File: DynamicProgramming/PalindromePartitioningII.py
class Solution():
    def __init__(self):
        pass

    def min_cut(self, str):
        length = len(str)
        is_pal_matrix = [[0] * length for _ in range(length)]
        dp = [length-1-i for i in range(length)] + [-1]

        for i in range(length-1, -1, -1):
            for j in range(i, length):
                if str[i] == str[j] and (j-i <= 1 or is_pal_matrix[i+1][j-1]):
                    is_pal_matrix[i][j] = 1
                    dp[i] = min(dp[i], dp[j+1] + 1)
        print(dp)
        return dp[0]

solu = Solution()
str = "abcbabaa"
print(solu.min_cut(str))
